source/common/router/rds_impl.cc

```diff
@@ -119,5 +125,45 @@ void RdsRouteConfigProviderImpl::registerInitTarget(Init::Manager& init_manager)
   init_manager.registerTarget(*this);
 }
 
+HttpRouteManagerImpl::HttpRouteManagerImpl(Runtime::Loader& runtime, Event::Dispatcher& dispatcher,
+                                           Runtime::RandomGenerator& random,
+                                           const LocalInfo::LocalInfo& local_info,
+                                           ThreadLocal::SlotAllocator& tls)
+    : runtime_(runtime), dispatcher_(dispatcher), random_(random), local_info_(local_info),
+      tls_(tls) {}
+
+std::vector<Router::RouteConfigProviderSharedPtr> HttpRouteManagerImpl::routeConfigProviders() {
+  std::vector<Router::RouteConfigProviderSharedPtr> ret;
+  ret.reserve(route_config_providers_.size());
+  for (const auto& element : route_config_providers_) {
+    ret.push_back(element.second.lock());
```

Comment:
I think I wasn't clear with my comment. I understand what lock does, and that it has nothing to do with threading. I was originally concerned that in the envoy threading model it would be possible to get to a state where the RouteConfigObject were destroyed but the weak ptr was still present in the Manager. I now understand that is not possible.

@htuch The pointer is weak because we want don't want the manager to own the RouteConfigProvider, we only want it to be able to dole out share ptrs to the HttpConnectionManagers that create RouteConfigProviders.